Various improvements

salman-farooq-sh opened the pull request:
Change interval_sec to interval so that interval can be configured to be less than a second but more than zero.

Allow configuring each notification's timeout/expiry time and enable/disable them.

Allow running custom commands on battery state changes.

Use `$SHELL` instead of `sh` and only use `sh` as a fallback when `$SHELL` isn't set.

Resolves https://github.com/cdown/battery-notify/issues/75.

salman-farooq-sh committed:
Remove interval_sec and add interval so that interval can be configured to be less than a second.
Add an option to disable state change notifications.
Allow configuring each notification's timeout/expiry time.

salman-farooq-sh committed:
Allow running custom commands on battery state changes.

Use $SHELL instead of sh.

cdown commented on README.md:
```diff
@@ -69,6 +82,18 @@ warn_on_mons_with_no_ac = 2
+# Negative values: Expire according to server default.
+sleep_pct_notif_timeout = 0
+low_pct_notif_timeout = 0
+state_notif_timeout = -1
```
Here and elsewhere, instead of using values like 0 (disable), -1 (default behavior), or 0 (no timeout), serialise as tagged unions to make the intention explicit.

cdown replied:
(`warn_on_mons_with_no_ac` is already violating this, but let's not add more)

salman-farooq-sh replied:
I guess that is better UX wise, this was just how the underlying notification library did it. Still the user will need to consult the documentation anyways to know the exact keywords, no?

salman-farooq-sh replied:
Just saw your second comment just now (after refreshing the page). So basically 0 (never), negative (server default), and positive (no timeout) is only for the notification timeouts.

cdown commented on README.md:
```diff
@@ -42,19 +42,32 @@ on first run, this will be populated with a basic config if it doesn't exist.
+# $SHELL -c <your-command>
+#
+# Using "true" like this is a no-op for most shells and the default.
+battery_state_charging_command = "true"
```
Use `Option<T>` instead of `true`.

salman-farooq-sh replied:
The true is a shell command/builtin that basically does nothing so I set that as the default.

cdown commented on README.md:
```diff
@@ -42,19 +42,32 @@ on first run, this will be populated with a basic config if it doesn't exist.
-# How often to check battery status, in seconds.
-interval_secs = 30
+# How often to check battery status, in milliseconds.
+interval = 30000
```
Removing the unit is confusing. Instead, let's use `#[serde_as(as = "humantime_serde::Duration")]`.

salman-farooq-sh replied:
But it will be a breaking change anyway, right? Because we can't leave the _secs suffix then. The user will have to consult the docs anyway as mentioned in another reply so this is okay being left like this in my opinion. Maybe this can be done some time later other than in this PR.

salman-farooq-sh replied:
Milliseconds are pretty standard for specifying time. More like a default I feel really unless mentioned explicitly (like *_secs).

salman-farooq-sh replied:
I don't think many people are going to need values large enough for this to justify being able to set something like "2h" instead. A minutes at most make sense for this config.

cdown replied:
The changes needed would be:

- If interval_secs is present, populate interval wtih it and print a deprecation message
- If interval is present, parse it with humantime_serde::Duration or similar
- If both are present, exit

In any case, the units should always be present. Over the years while developing systemd we've had countless problems with directives like `*USec`, `*Sec` and the like. The best way to avoid that in future is to always require units.

salman-farooq-sh replied:
I see, I have to agree with you then on units being always present, with your level of experience.

Regarding breaking-change right away vs. deprecation message first, what will we do about the rest of the changes which might also be breaking?

salman-farooq-sh replied:
What if we simply call it `interval_ms` instead? Similarly for notifications.

cdown replied:
I'm okay with either not changing it, or using interval with humantime parsing, but let's not break backwards compat just for that when I don't think there's any demand for ms granularity.

salman-farooq-sh replied:
I imagine using floats for interval_secs is out of question? We can convert to ms/us while parsing the config. Otherwise I will try doing humantime with depraction notice, and if I fail we can let it be as it is. Thanks.

cdown commented on README.md:
```diff
@@ -69,6 +82,18 @@ warn_on_mons_with_no_ac = 2
 bluetooth_low_pct = 40
+
+# Set to false to disable state change notifications e.g. when charging, discharging, reaching the threshold
+state_notif_enabled = true
```
This should be an enum showing which state change notifications should be shown. See below for more about that.

salman-farooq-sh replied:
Like a single config keyword that is a list/array?

cdown replied:
Ditto directly above :-)

salman-farooq-sh replied:
Marking this as resolved because I think it might already be done, see latest comment at the bottom.

cdown commented on src/main.rs:
```diff
@@ -99,28 +128,52 @@ fn main() -> Result<()> {
+        if battery_state_name != last_battery_state_name {
+            last_battery_state_name = battery_state_name.clone();
+
+            match battery_state_name.as_str() {
```
Any time you write something like this, what you wanted was an enum :-) Among other things:

1. Enums ensure all possible states are predefined, and the compiler enforces exhaustive matching. In this case, for example, mistyped states in string matching (e.g., `"Chargin"`) would not be caught at compile time.
2. Adding new states or renaming existing ones is safer and more straightforward with enums, as the compiler ensures all usage is updated and is exhaustive.

cdown commented on README.md:
```diff
@@ -42,19 +42,32 @@ on first run, this will be populated with a basic config if it doesn't exist.
+#
+# Using "true" like this is a no-op for most shells and the default.
+battery_state_charging_command = "true"
+battery_state_discharging_command = "true"
```
Since this is now getting a lot of values, let's update to:

Use sections, then you can have thing like:

```
[state_notif]
full_cmd = ...
...
```

...and not have to prefix everything

cdown commented on src/main.rs:
```diff
@@ -99,28 +128,52 @@ fn main() -> Result<()> {
+            low_notif.show(
+                "Battery low".to_string(),
+                Urgency::Critical,
+                cfg.low_pct_notif_timeout,
```
If you use an enum, you can map all of this information: the string, the criticality, and the timeout, and just do `low_notif.show()`, and similarly elsewhere. Then we don't have to reimplement for every type of notification.

salman-farooq-sh replied:
Can you explain a bit more please? Thanks.

cdown replied:
I'm suggesting using a tagged union and struct to store some of the repeated info. :-)

But I must also confess I don't understand the point of having a different timeout for every single notification type. What's the use case? Urgency::Critical notifications almost always are not dismissed anyway.

Overriding this by the application should generally not be done, the user sets their own preferences in the notification daemon. So unless there is a compelling reason, the timeout stuff should be configured there, not here.

salman-farooq-sh replied:
> I'm suggesting using a tagged union and struct to store some of the repeated info. :-)

I think I might have done just that already, this is a bit outdated. If not, let me know what improvement can be made. Thanks.

> But I must also confess I don't understand the point of having a different timeout for every single notification type. What's the use case? Urgency::Critical notifications almost always are not dismissed anyway.

I am using [mako](https://github.com/emersion/mako), a very popular server, and it doesn't have per-urgency timeouts.

> Overriding this by the application should generally not be done, the user sets their own preferences in the notification daemon. So unless there is a compelling reason, the timeout stuff should be configured there, not here.

So this is now more of allowing the user to enable/disable individual notifications, while having some extra flexibility while at it. `notify-rust` uses the same integer timeout parameter for making notifications persistent slash default timeout slash configurable timeout.

Wouldn't applications be disallowed to select this stuff by their choosing if it wasn't meant to be so? Also, it gets a bit tedious at best and very hacky at worst to configure timeout for each notification separately in the notification server.

cdown commented on README.md:
```diff
@@ -42,19 +42,32 @@ on first run, this will be populated with a basic config if it doesn't exist.
+interval = 30000
 
-# At what percentage of battery capacity to notify about low battery.
+# At what percentage of battery capacity to notify about low battery, set to 0 to disable.
```
Considering #73, instead of making low_pct/sleep_pct more nuanced, let's make it generic:

1. Have an enum that represents all notification states
2. Allow the user to define what happens when those states are triggered (can be configured as a table in the TOML)
3. Refuse to run if we see a state that's not known

salman-farooq-sh replied:
Like provide full granularity of enabling/disabling every notification type we send? Considering every battery state as a separate notification type? I am not sure I completely understand what you said.

salman-farooq-sh replied:
1. ~~Have an enum that represents all notification states~~
1. ~~Allow the user to define what happens when those states are triggered (can be configured as a table in the TOML)~~
1. Refuse to run if we see a state that's not known

cdown replied:
This was marked as resolved, but it doesn't use the enum, right?

salman-farooq-sh replied:
It uses a struct now, please check the latest commit. I think only the refusing to run part is left now.

cdown reviewed (commented):
Thanks! Some review comments.

cdown commented:
(Also, each logical change should be a separate commit.)

salman-farooq-sh committed:
new config

cdown commented:
Thanks for working on this, I'll look when I get a chance with Christmas. Have a good end of your year :-)

salman-farooq-sh committed:
Merge branch 'master' of https://github.com/cdown/battery-notify

salman-farooq-sh commented:
My config being used with this branch right now:

```toml
interval = 1000
low_pct = 25
sleep_pct = 5
bluetooth_low_pct = 20
monitors_with_no_ac = 2

[events]
low = "idle-reset"
sleep = "systemctl suspend"
charging = "idle-reset"
discharging = "idle-reset"
not_charging = "idle-reset"
full = "idle-reset"
unknown = "idle-reset"
at_threshold = "idle-reset"

[notifications]
low = "persistent"
sleep = "persistent"
bluetooth_low = "persistent"
monitors_with_no_ac = "persistent"
charging = "disabled"
discharging = "disabled"
not_charging = "disabled"
full = "disabled"
unknown = "disabled"
at_threshold = "disabled"
```

salman-farooq-sh commented:
@cdown Hi, let's start on this again. Can you give a new review? Thanks.

cdown commented:
Hey, I'm currently in a remote area of Costa Rica with no consistent access
to the internet. I'll be back in a few weeks. Thanks for your patience :⁠-⁠)

On Sun, 12 Jan 2025, 11:40 Salman Farooq, ***@***.***> wrote:

> @cdown <https://github.com/cdown> Hi, let's start on this again. Can you
> give a new review? Thanks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cdown/battery-notify/pull/78#issuecomment-2585846279>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFBJN7RRH4YDLDFKK426B32KKSJBAVCNFSM6AAAAABTQE6Q4KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOBVHA2DMMRXHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

salman-farooq-sh commented:
Oh, have a nice trip!

salman-farooq-sh committed:
config: Extract Config into own module

Pure refactor with no behavioral changes.

salman-farooq-sh committed:
config: Replace interval_secs with interval (humantime), rename warn_on_mons_with_no_ac to monitors_with_no_ac

Add humantime dependency for human-readable interval parsing (e.g. "30s", "500ms").
Both old keys are deprecated with warnings; specifying both old and new exits
with an error.

salman-farooq-sh committed:
notify: Add timeout parameter and ConfiguredNotification

SingleNotification::show() now accepts a timeout (i32) forwarded to the
notification daemon. Introduce ConfiguredNotification to bundle a fixed
summary, urgency, and timeout for repeated use.

salman-farooq-sh committed:
config/main: Add [notifications] and [events] sections, state change handling, $SHELL support

Add per-notification configurability with [notifications] section — each
notification can be set to persistent, server-default, disabled, or a
specific timeout in milliseconds.

Add [events] section for custom commands on any battery state change or
threshold crossing. Commands run via $SHELL -c (falls back to sh).

Deprecate sleep_command in favor of events.sleep.

Refactor state change handling into handle_state_change() for deduplication.

salman-farooq-sh committed:
docs: Update README for new configuration options

Document humantime interval, [events] section, [notifications] section,
and add a migration guide for deprecated keys.
